### PR TITLE
Add shard/device env vars for screenshots

### LINKED ISSUE #2164: Screenshot function only saves a single screenshot when run on multiple devices
### Is there an existing issue for this?

- [X] I have searched the existing issues and didn't find mine.

### Steps to reproduce

1. Create a flow file that takes a screenshot
```
- takeScreenshot:
    path: LoginScreen
```
2. Run the flow on multiple devices or use shard-all `maestro test --shard-all 2 .`

### Actual results

You will only have 1 screenshot even though you ran this on multiple devices (they override each other)

### Expected results

You should get 1 screenshot per device saved to disk. Ideally we would be provided some environment variables that can be used when setting the path of the screenshot to avoid clashing between the runs and be able to attribute the screenshots to the respective devices. For example `MAESTRO_SHARD_ID` and `MAESTRO_DEVICE_UDID`.

### About app

This is an open source app, available at https://github.com/wikimedia/wikipedia-ios

### About environment

Apple M1 Max Macbook Pro 16

### Logs

<details>
<summary>Logs</summary>

```
<!-- Replace this line with your logs. *DO NOT* remove the backticks! -->
```

</details>

### Maestro version

1.39.2

### How did you install Maestro?

install script (https://get.maestro.mobile.dev)

### Anything else?

Tagging @Fishbowler 
[Link to our Slack conversation ](https://mobile-dev-inc.slack.com/archives/C041FU72T54/p1732811004883429)

### PR DESCRIPTION
**Pull request by ImL1s**
## Summary

Adds `MAESTRO_DEVICE_UDID`, `MAESTRO_SHARD_ID`, and `MAESTRO_SHARD_INDEX` as default flow environment variables to support unique screenshot filenames in multi-shard/multi-device scenarios.

Fixes #2164

## Usage

```yaml
# Use in screenshot command to avoid filename conflicts
- takeScreenshot: "screenshot_${MAESTRO_SHARD_ID}_${MAESTRO_DEVICE_UDID}.png"

# Or use 0-based index for programmatic naming
- takeScreenshot: "shard_${MAESTRO_SHARD_INDEX}/test_${MAESTRO_DEVICE_UDID}.png"

# Safe filename with sanitized device ID
- evalScript: |
    const safeDeviceId = MAESTRO_DEVICE_UDID.replace(/[^a-zA-Z0-9-]/g, '_')
    output.filename = `screenshot_${safeDeviceId}.png`
- takeScreenshot: ${output.filename}
```

**Available environment variables:**
- `MAESTRO_DEVICE_UDID`: Device unique identifier (e.g., "iPhone-15-Pro", "emulator-5554")
- `MAESTRO_SHARD_ID`: 1-based shard number (e.g., "1", "2", "3") - user-friendly
- `MAESTRO_SHARD_INDEX`: 0-based shard index (e.g., "0", "1", "2") - programmer-friendly

## Changes

- Add `deviceId` and `shardIndex` parameters to `Env.withDefaultEnvVars()` in `maestro-orchestra-models`
- Thread device/shard context through CLI test runners:
  - `TestCommand`: Pass `deviceId` to single and multiple flow runs
  - `RecordCommand`: Pass `deviceId` to test runner (fixed to use local variable consistently)
  - `TestRunner`: Add `deviceId` parameter to `runSingle()` and `runContinuous()`
  - `TestSuiteInteractor`: Add both `deviceId` and `shardIndex` for batch execution
- Add MCP tool support:
  - `RunFlowTool`: Propagate `deviceId` to flow environment
  - `RunFlowFilesTool`: Propagate `deviceId` to flow environment
- Add comprehensive unit tests in `EnvTest.kt`
- Add integration test `134_shard_device_env_vars.yaml` demonstrating usage

## Testing

```bash
# Unit tests
./gradlew :maestro-orchestra-models:test

# Integration tests (including new Case 134)
./gradlew :maestro-test:test

# Build check
./gradlew :maestro-cli:compileKotlin

# E2E test with sharding (manual verification)
./maestro test --platform ios --shard-all 2 /tmp/maestro_shard_screenshot.yaml --test-output-dir /tmp/maestro-shard-output-ios
```

## Notes

- **Why two shard variables?**
  - `MAESTRO_SHARD_ID` (1-based): Better for user-facing output like "Shard 1 of 3"
  - `MAESTRO_SHARD_INDEX` (0-based): Better for array indexing and programmatic use
  
- **Override behavior**: New env vars override existing values if present

- **MCP support**: MCP tools receive `deviceId` but not `shardIndex` (MCP typically runs single flows, not sharded batches)

- **RecordCommand**: Only receives `deviceId`, not `shardIndex` (recording mode doesn't support sharding)

- **⚠️ File path safety**: `MAESTRO_DEVICE_UDID` may contain spaces or special characters (e.g., "iPhone 15 Pro", "Pixel 7/Simulator"). When using in file paths:
  - **Recommended**: Use `evalScript` to sanitize the device ID before using in filenames (see Usage examples above)
  - **Alternative**: Ensure your file system and tools handle special characters correctly
  - Device IDs are typically safe for directory names but may cause issues in certain contexts

- **Variable availability**: These variables are only set when the corresponding context is available:
  - `MAESTRO_DEVICE_UDID`: Set when deviceId is provided to the flow
  - `MAESTRO_SHARD_ID/INDEX`: Only set during `--shard-all` execution

### CODE REVIEW

File: maestro-orchestra-models/src/test/kotlin/maestro/orchestra/util/EnvTest.kt [hunk 1]
```diff
@@ -28,6 +28,30 @@ class EnvTest {
...
+    fun `withDefaultEnvVars should override shard and device values`() {
+        val env = mapOf(
+            "MAESTRO_DEVICE_UDID" to "old-device",
+            "MAESTRO_SHARD_ID" to "99",
+            "MAESTRO_SHARD_INDEX" to "98",
+        ).withDefaultEnvVars(deviceId = "device-456", shardIndex = 0)
+        assertThat(env["MAESTRO_DEVICE_UDID"]).isEqualTo("device-456")
+        assertThat(env["MAESTRO_SHARD_ID"]).isEqualTo("1")
+        assertThat(env["MAESTRO_SHARD_INDEX"]).isEqualTo("0")
+    }
```
[the remark this answers is not in the document]
Comment (ImL1s) [reply maestro-orchestra-models/src/test/kotlin/maestro/orchestra/util/EnvTest.kt]: Done! Added 4 edge case tests:
- Empty string deviceId
- Blank deviceId
- Only deviceId without shardIndex
- Only shardIndex without deviceId

All tests pass ✓

File: maestro-cli/src/main/java/maestro/cli/command/RecordCommand.kt [hunk 2]
```diff
@@ -150,7 +150,16 @@ class RecordCommand : Callable<Int> {
...
-                        TestRunner.runSingle(maestro, device, flowFile, env, resultView, path, testOutputDir = null)
+                        TestRunner.runSingle(
+                            maestro,
+                            device,
+                            flowFile,
+                            env,
+                            resultView,
+                            path,
+                            testOutputDir = null,
+                            deviceId = parent?.deviceId,
```
[the remark this answers is not in the document]
Comment (ImL1s) [reply maestro-cli/src/main/java/maestro/cli/command/RecordCommand.kt]: Done! Not only using the local variable, but improved it to use `device.instanceId` instead of `deviceId` (CLI parameter).

This ensures MAESTRO_DEVICE_UDID is populated with the actual selected device ID, even when `--device` flag is not specified.

Changed:
```kotlin
deviceId = device.instanceId,  // Previously: deviceId = parent?.deviceId
```

**Comment by Fl0p** [comment]
Nice idea. i use this as workaround
```
# preconditions
- assertTrue: ${!!SCREENSHOT_NAME}

- evalScript: ${output.screenshot = output.screenshot || {}}
- evalScript: ${output.screenshot.counter = output.screenshot.counter || 1}
- evalScript: ${output.screenshot.prefix = ("000" + output.screenshot.counter).slice(-4)}

# take screenshot
- takeScreenshot: ${maestro.flowName}/${output.screenshot.prefix}_${SCREENSHOT_NAME}
- evalScript: ${output.screenshot.counter = output.screenshot.counter + 1}

```

File: maestro-test/src/test/resources/134_shard_device_env_vars.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+appId: com.example.app
+---
+# Test MAESTRO_DEVICE_UDID, MAESTRO_SHARD_ID, and MAESTRO_SHARD_INDEX env vars
+- launchApp
+- takeScreenshot: ${MAESTRO_FILENAME}_${MAESTRO_DEVICE_UDID}_shard${MAESTRO_SHARD_ID}
```
Comment (Fishbowler) [inline maestro-test/src/test/resources/134_shard_device_env_vars.yaml]: Despite the comment, `MAESTRO_SHARD_INDEX` isn't tested here

**Review by Fishbowler — COMMENTED** [review]
This looks tidy! One tiny nitpick...

File: maestro-test/src/test/resources/134_shard_device_env_vars.yaml (hunk 3 above)
Comment (Fishbowler) [reply maestro-test/src/test/resources/134_shard_device_env_vars.yaml]: You've accidentally committed some extra files in that commit, I think

**Comment by ImL1s** [comment]
You're absolutely right! 😅 I clearly needed a human to help me count to 136 correctly.

Third time's the charm - I promise the test numbers are correct now:
- Case 134: Take screenshot with path (upstream's)
- Case 135: Screen recording with path (upstream's)  
- Case 136: Shard and device env vars (mine)

Thanks for your patience with my arithmetic! 🙏

**Comment by ImL1s** [comment]
That's interesting! 🤔 According to the code, when `shardIndex` is `null`, both `MAESTRO_SHARD_ID` and `MAESTRO_SHARD_INDEX` should be undefined (lines 45-48 in Env.kt).

If `MAESTRO_SHARD_ID = 1` but `MAESTRO_SHARD_INDEX = undefined`, that suggests `shardIndex` was passed as `0` somewhere but got lost for the INDEX variable. 

Could you share how you ran it? (command line flags, etc.) This will help me debug where the inconsistency is coming from.

**My intended behavior:**
- **With sharding** (`--shard-all 2`): All three vars should be set
- **Without sharding** (regular `maestro test`): Only `DEVICE_UDID` should be set, both shard vars undefined

The current state (ID=1, INDEX=undefined) shouldn't happen. 🐛

**Comment by ImL1s** [comment]
## Update: Reproduced and confirmed the behavior! 🔬

I ran a test locally with 2 flows (no sharding flags):

```bash
MAESTRO_DRIVER_STARTUP_TIMEOUT=120000 maestro test --platform ios flow1.yaml
```

**From the debug log**, I can see exactly what env vars are being set:
```
DefineVariablesCommand(env={
  MAESTRO_DRIVER_STARTUP_TIMEOUT=120000,  ← injected from MY shell
  MAESTRO_FILENAME=flow1, 
  MAESTRO_DEVICE_UDID=7C25D515-FF99-43C2-A2AD-136A1A0643E0
})
```

**Key observations:**
- ✅ `MAESTRO_DEVICE_UDID` is set correctly by my code
- ❌ `MAESTRO_SHARD_ID` is **NOT** set (because `shardIndex=null` when not sharding)
- ❌ `MAESTRO_SHARD_INDEX` is **NOT** set
- 🔍 `MAESTRO_DRIVER_STARTUP_TIMEOUT=120000` was **injected from my shell environment** by `withInjectedShellEnvVars()`

This confirms the code is working correctly! The shell env vars are being passed through, which is how `MAESTRO_SHARD_ID=1` must have appeared in your test.

**Conclusion:** If you're seeing `MAESTRO_SHARD_ID=1`, it's being injected from your shell/CI environment via `withInjectedShellEnvVars()`. Could you double-check with `env | grep MAESTRO`? 🙏

**Comment by Fishbowler** [comment]
That's not it. There's no way I could have a brand new env var already present in my env. I'm running locally, not in CI or Docker. I'm using VSCode, so no overblown env management going on.

**Comment by Fishbowler** [comment]
I haven't rebuilt and can't currently reproduce the problem either. Maybe we should revert the last couple of commits, and suggest that maybe we're overworking for something that might or might not be real, and might or might not be trivial anyway.

**Comment by ImL1s** [comment]
Great question\! 🤔

**Scenario**: Flow uses `${MAESTRO_SHARD_ID}` in screenshot filename. CI runs with sharding → works. Developer opens in Maestro Studio to debug → what happens?

**Current behavior (after my fix)**: Variables are `undefined` when not sharding.

**Options**:
1. **Keep undefined** - Clean semantics. Flows need conditional checks like:
   ```js
   ${typeof MAESTRO_SHARD_ID \!== 'undefined' ? MAESTRO_SHARD_ID : 'local'}
   ```
2. **Provide defaults** - Always set `SHARD_ID=1, SHARD_INDEX=0` when not sharding. Better DX but semantically misleading ("shard 1" when there's no sharding?)

I lean toward option 1 because:
- Variables should reflect actual state
- Users who need shard info in filenames probably want it to be accurate, not a fake default
- The conditional pattern is common in JS (`|| 'fallback'`)

But if DX is a bigger concern, we could add defaults. What do you think?

**Review by Fishbowler — APPROVED** [review]
Local testing of this went great:
- Values are correct across devices
- Values remain stable between runs (i.e. the UDID doesn't change)
- Defaults work great with `--shard-split 1` or omitted
